cloud/libvirt/actuators/machine/actuator.go

```diff
@@ -68,5 +69,98 @@ func (a *Actuator) Update(cluster *clusterv1.Cluster, machine *clusterv1.Machine
 // Exists test for the existance of a machine and is invoked by the Machine Controller
 func (a *Actuator) Exists(cluster *clusterv1.Cluster, machine *clusterv1.Machine) (bool, error) {
 	glog.Infof("Checking if machine %v for cluster %v exists.", machine.Name, cluster.Name)
-	return libvirtutils.DomainExists(machine)
+
+	// decode config
+	machineProviderConfig, err := machineProviderConfigFromClusterAPIMachineSpec(&machine.Spec)
+	if err != nil {
+		return false, fmt.Errorf("error getting machineProviderConfig from spec: %v", err)
+	}
+
+	// build libvirtd client
+	client, err := libvirtutils.BuildClient(machineProviderConfig.URI)
+	if err != nil {
+		return false, fmt.Errorf("Failed to build libvirt client: %s", err)
+	}
+	return libvirtutils.DomainExists(machine.Name, client)
+}
+
+// CreateVolumeAndMachine creates a volume and domain which consumes the former one
+func createVolumeAndDomain(machine *clusterv1.Machine, offset int) error {
+	// decode config
+	machineProviderConfig, err := machineProviderConfigFromClusterAPIMachineSpec(&machine.Spec)
+	if err != nil {
+		return fmt.Errorf("error getting machineProviderConfig from spec: %v", err)
+	}
+
+	// build libvirtd client
+	client, err := libvirtutils.BuildClient(machineProviderConfig.URI)
+	if err != nil {
+		return fmt.Errorf("Failed to build libvirt client: %s", err)
+	}
+
+	// TODO(alberto) create struct and function for converting machineconfig->libvirtConfig
+	name := machine.Name
+	baseVolume := machineProviderConfig.Volume.BaseVolumeID
+	pool := machineProviderConfig.Volume.PoolName
+	ignKey := machineProviderConfig.IgnKey
+	networkInterfaceName := machineProviderConfig.NetworkInterfaceName
+	networkInterfaceAddress := machineProviderConfig.NetworkInterfaceAddress
+	autostart := machineProviderConfig.Autostart
+	memory := machineProviderConfig.DomainMemory
+	vcpu := machineProviderConfig.DomainVcpu
+
+	// Create volume
+	if err := libvirtutils.CreateVolume(name, pool, baseVolume, "", "qcow2", client); err != nil {
+		return fmt.Errorf("error creating volume: %v", err)
+	}
+
+	// Create domain
+	if err = libvirtutils.CreateDomain(name, ignKey, name, name, networkInterfaceName, networkInterfaceAddress, autostart, memory, vcpu, offset, client); err != nil {
+		return fmt.Errorf("error creating domain: %v", err)
+	}
+	return nil
+}
+
+// deleteVolumeAndDomain deletes a domain and its referenced volume
+func deleteVolumeAndDomain(machine *clusterv1.Machine) error {
```

Comment:
nit: split the functionality into discrete steps.

Reply:
it's already 
```
	// delete domain
	if err := libvirtutils.DeleteDomain(name, client); err != nil {
		return fmt.Errorf("error deleting domain: %v", err)
	}
 	// delete volume
	if err := libvirtutils.DeleteVolume(name, client); err != nil {
		return fmt.Errorf("error deleting volume: %v", err)
	}
```

Reply:
I was thinking of _it_ (the func) doing one thing only, and one thing really well. This does two things. If the volume failed to delete I could subsequently call deleteDomain - here it tries to do too much and as a caller I can't distinguish between which failed.